### PR TITLE
feat(CORE/Authserver): Removed sha_pass_hash from DB and core.

### LINKED ISSUE #3352: Security Vulnerability (Fix included) [$200]
https://github.com/cmangos/issues/issues/1826
https://github.com/TrinityCore/TrinityCore/issues/25157
https://github.com/TrinityCore/TrinityCore/issues/22913
same shit here boys 

<bountysource-plugin>

---
There is a **[$200 open bounty](https://www.bountysource.com/issues/92463000-security-vulnerability-fix-included?utm_campaign=plugin&utm_content=tracker%2F40032087&utm_medium=issues&utm_source=github)** on this issue. Add to the bounty at [Bountysource](https://www.bountysource.com/?utm_campaign=plugin&utm_content=tracker%2F40032087&utm_medium=issues&utm_source=github).
</bountysource-plugin>

### PR DESCRIPTION
**Pull request by UltraNix**
Ported from TrinityCore and adapted to AzerothCore.
Fixed #3352.

## Changes Proposed:
-  Deprecation of sha_pass_hash.

## Issues Addressed:
- Closes #3352

## Tests Performed:
- Tested ingame. Works okay with old and newly created accounts. Password changing works as well.
-

## Known Issues and TODO List:

- [ ]
- [ ] 

## Target Branch(es):
- [x] Master

 
## How to Test AzerothCore PRs
 
When a PR is ready to be tested, it will be marked as **[WAITING TO BE TESTED]**.

You can help by testing PRs and writing your feedback here in the PR's page on GitHub. Follow the instructions here:

http://www.azerothcore.org/wiki/How-to-test-a-PR

### CODE REVIEW

File: data/sql/updates/pending_db_auth/rev_1615629613255169700.sql [hunk 1]
```diff
@@ -0,0 +1,27 @@
...
+
+-- update `account` structure
+-- sha_pass_hash/s/v kept around for now, for backwards compatibility
+ALTER TABLE `account`
+    DROP COLUMN `sessionkey`,
+    ADD COLUMN `salt` BINARY(32) AFTER `username`,
+    ADD COLUMN `verifier` BINARY(32) AFTER `salt`,
+    ADD COLUMN `session_key` BINARY(40) AFTER `verifier`,
+    MODIFY COLUMN `s` VARCHAR(64) NOT NULL DEFAULT 'dummy value, use `salt` instead',
+    MODIFY COLUMN `v` VARCHAR(64) NOT NULL DEFAULT 'dummy value, use `verifier` instead';
```
Comment (FrancescoBorzi) [inline data/sql/updates/pending_db_auth/rev_1615629613255169700.sql:11]: do we need to keep `s` and `v` ?

**Comment by FrancescoBorzi** [comment]
thanks for your contribution @UltraNix , please check the build that is failing

**Comment by UltraNix** [comment]
Okay. I will fix them all later.

**Comment by asfo** [comment]
Maybe before this kind of changes happens, there should be a "message/notification" that every webpage for register accounts in AzerothCore are going to change and this is a "breaking change" for websites, because this could lead to a lot of issues in a lot of servers (I know the responsability is from the web devs or stuff like that), just a comment because of this:

image

That states "kept around" and then "removed".

Or at least update the documentation/wiki to know the new "register process"? Or not sure how this is going to work.

**Comment by UltraNix** [comment]
> Maybe before this kind of changes happens, there should be a "message/notification" that every webpage for register accounts in AzerothCore are going to change and this is a "breaking change" for websites, because this could lead to a lot of issues in a lot of servers (I know the responsability is from the web devs or stuff like that), just a comment because of this:
> 
> image
> 
> That states "kept around" and then "removed".
> 
> Or at least update the documentation/wiki to know the new "register process"? Or not sure how this is going to work.

This is an old comment, as the sha_pass_hash is completely removed.

**Comment by UltraNix** [comment]
Ready to merge.

**Comment by Kitzunu** [comment]
@Viste

File: src/common/Cryptography/BigNumber.cpp [hunk 2]
```diff
@@ -99,7 +111,7 @@ BigNumber BigNumber::operator/=(BigNumber const& bn)
     BN_CTX* bnctx;
 
     bnctx = BN_CTX_new();
-    BN_div(_bn, nullptr, _bn, bn._bn, bnctx);
+    BN_div(_bn, NULL, _bn, bn._bn, bnctx);
```
Comment (Winfidonarleyan) [inline src/common/Cryptography/BigNumber.cpp:114]: ?

**Comment by FrancescoBorzi** [comment]
@azerothcore/developers @azerothcore/testers @pklloveyou @mpfans please check this

**Comment by pklloveyou** [comment]
I'm coming.

**Comment by pklloveyou** [comment]
What should I test, After pr now I can log in to the game

**Comment by pklloveyou** [comment]
account set pass admins 123123
Server crash
```BigNumber.cpp:142 in BigNumber::GetBytes ASSERTION FAILED:
 res > 0
Buffer of size 0 is too small to hold bignum with 0 bytes.```

**Comment by akrom23** [comment]
@pklloveyou What is your OpenSSL version?

**Comment by UltraNix** [comment]
@pklloveyou What is your OpenSSL version?

**Comment by pklloveyou** [comment]
OpenSSL 1.1.1g
I'll try to rebuild it

**Comment by UltraNix** [comment]
I see it - crash is caused by corrupted data in warden checks with MEM_CHECK type. There are 6 checks with null data - is it intended?

**Comment by pklloveyou** [comment]
Warden.Enabled = 0 So it doesn't crash anymore

**Comment by UltraNix** [comment]
But this is not the solution.

**Comment by pklloveyou** [comment]
yes，Sorry, the translation is not accurate, I don't mean that

**Comment by UltraNix** [comment]
Two solutions:

1. Fix those 6 warden checks (right now they are useless without any data).
2. Remove asserts so the core won't crash  (nothing breaks here).

**Comment by FrancescoBorzi** [comment]
> Fix those 6 warden checks (right now they are useless without any data).

@UltraNix can you please fix them?

**Comment by UltraNix** [comment]
Yes - by porting them from TC.

**Comment by FrancescoBorzi** [comment]
> Yes - by porting them from TC.
> Should I make a new PR with it or just add it to this one?

thanks.
smaller PRs are always better, so it can be separately reviewed and tested

**Comment by pklloveyou** [comment]
image
Is this the right way to write code?

**Comment by UltraNix** [comment]
Looks okay.

**Comment by pklloveyou** [comment]
I will use this PR on the server tomorrow

**Comment by Helias** [comment]
solve merge conflicts pls

**Comment by UltraNix** [comment]
Updated.

**Comment by FrancescoBorzi** [comment]
> I will use this PR on the server tomorrow

@pklloveyou  any news?

**Comment by pklloveyou** [comment]
Can be merged

**Comment by FrancescoBorzi** [comment]
thank you! @pklloveyou

**Review by FrancescoBorzi — APPROVED** [review]

**Pull request merged**